### PR TITLE
HPCC-15717 Support paging in retrieving permissions

### PR DESCRIPTION
**Pull request by wangkx**
A new ws_access method 'ResourcePermissionQuery' is added to
support paging in retrieving permissions. The new method also
allows filtering the permissions by account type and sorting
the permissions by permission names or account types.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by wangkx** [comment]
@RussWhitehead please review.

File: esp/services/ws_access/ws_accessService.cpp [hunk 1]
```diff
@@ -2192,6 +2183,75 @@ bool Cws_accessEx::onResourcePermissions(IEspContext &context, IEspResourcePermi
...
+        __int64 pageStartFrom = 0;
+        unsigned pageSize = 100;
+        if (!req.getPageSize_isNull())
+            pageSize = req.getPageSize();
+        if (!req.getPageStartFrom_isNull())
+            pageStartFrom = req.getPageStartFrom();
+
+        ResourcePermissionField sortOrder[2] = {RPFName, RPFterm};
+        CResourcePermissionSortBy sortBy = req.getSortBy();
+        switch (sortBy)
```
Comment (RussWhitehead) [inline esp/services/ws_access/ws_accessService.cpp]: It's not clear to me why this was done with a switch statement

File: esp/scm/ws_access.ecm [hunk 2]
```diff
@@ -995,6 +1032,7 @@ ESPservice [version("1.10"), auth_feature("NONE"), exceptions_inline("./smc_xslt
     ESPmethod GroupQuery(GroupQueryRequest, GroupQueryResponse);
     ESPmethod GroupMemberQuery(GroupMemberQueryRequest, GroupMemberQueryResponse);
     ESPmethod ResourceQuery(ResourceQueryRequest, ResourceQueryResponse);
+    ESPmethod ResourcePermissionQuery(ResourcePermissionQueryRequest, ResourcePermissionQueryResponse);
```
Comment (RussWhitehead) [inline esp/scm/ws_access.ecm]: do you need to version protect this with "min_ver" ?

**Review by RussWhitehead — CHANGES_REQUESTED** [review]
A few minor comments/questions

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 3]
```diff
@@ -3618,6 +3627,109 @@ class CLdapClient : implements ILdapClient, public CInterface
...
+        getPermissionsArray(baseDN, type, namebuf.str(), permissions);
+
+        Owned<IPropertyTree> permissionTree = createPTree("Permissions");
+        ForEachItemIn(i, permissions)
+            addPermissionTree(permissions.item(i), accountType, permissionTree);
+        return permissionTree->getElements("*");
+    }
+
+    ISecItemIterator* getResourcePermissionsSorted(const char* name, enum ACCOUNT_TYPE_REQ accountType, const char* baseDN, const char* rtype,
+        const char* prefix, ResourcePermissionField* sortOrder, const unsigned pageStartFrom, const unsigned pageSize,
```
[the remark this answers is not in the document]
Comment (wangkx) [reply system/security/LdapSecurity/ldapconnection.cpp]: No. The unsigned is needed by the 'getElementsPaged()' inside the dautils.cpp.

File: esp/services/ws_access/ws_accessService.cpp [hunk 4]
```diff
@@ -2106,6 +2106,36 @@ bool Cws_accessEx::onResourceDelete(IEspContext &context, IEspResourceDeleteRequ
...
 }
 
+void Cws_accessEx::addResourcePermission(const char *name, int type, int allows, int denies, IArrayOf<IEspResourcePermission> &permissions)
+{
+    if (isEmptyString(name))
+        return;
+
+    StringBuffer nameIn = name;
+    Owned<IEspResourcePermission> permission = createResourcePermission();
+    permission->setAccount_name(nameIn.str());
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/services/ws_access/ws_accessService.cpp]: Will do

File: esp/services/ws_access/ws_accessService.cpp (hunk 1 above)
Comment (wangkx) [reply esp/services/ws_access/ws_accessService.cpp]: The switch was used in other places due to multiple sortby's. It should not be used here.

File: esp/scm/ws_access.ecm (hunk 2 above)
Comment (wangkx) [reply esp/scm/ws_access.ecm]: Will add...

**Comment by wangkx** [comment]
@RussWhitehead please review my changes.

**Comment by RussWhitehead** [comment]
@wangkx Looks good

**Comment by wangkx** [comment]
@richardkchapman this PR is ready to be merged.

**Pull request merged**